1.21.1 update

JayemCeekay opened the pull request:
Broke up UI texture atlases into singular sprite textures where needed.
Fixed several issues regarding changes to codecs in 1.21.1.
Changed the quest description markdown to rely mainly on HTML-like tags.
Fixed Neoforge module.

JayemCeekay committed:
1.21.1 update
broke up ui texture atlases into singular sprite textures where needed.
Fixed several issues regarding changes to codecs in 1.21.1.
Changed the quest description markdown to rely mainly on HTML-like tags.
Fixed neoforge module

MsRandom commented on common/src/main/java/earth/terrarium/heracles/api/client/WidgetUtils.java:
```diff
@@ -33,8 +33,8 @@ public final class WidgetUtils {
-        graphics.blitNineSliced(TEXTURE, x, y, 42, height, 3, 42, 42, 0, 0);
-        graphics.blitNineSliced(TEXTURE, x + 42, y, width - 42, height, 3, 86, 42, 42, 0);
+        graphics.blitSprite(ResourceLocation.fromNamespaceAndPath(Heracles.MOD_ID, "widgets/background"), x, y, 42, height);
+        graphics.blitSprite(ResourceLocation.fromNamespaceAndPath(Heracles.MOD_ID, "widgets/background2"), x + 42, y, width - 42, height);
```
These should ideally remain constants.

MsRandom replied:
Same with every instance of `ResourceLocation.fromNamespaceAndPath` used inline here

ThatGravyBoat replied:
Should also have a method in Heracles named `id` that does the `ResourceLocation.fromNamespaceAndPath(MOD_ID, ...)`

MsRandom commented on common/src/main/java/earth/terrarium/heracles/api/client/settings/tasks/BlockInteractTaskSettings.java:
```diff
@@ -32,7 +33,7 @@ public BlockInteractTask create(String id, @Nullable BlockInteractTask object, D
             data.get("block", RegistryValueSetting.BLOCK).orElse(getDefaultBlock(object)),
             Optionull.mapOrDefault(object, BlockInteractTask::state, BlockStatePredicate.ANY),
-            Optionull.mapOrDefault(object, BlockInteractTask::nbt, NbtPredicate.ANY)
+            Optionull.mapOrDefault(object, BlockInteractTask::nbt, new NbtPredicate(new CompoundTag()))
```
Can we move `new NbtPredicate(new CompoundTag())` to a static utility?

MsRandom commented on common/src/main/java/earth/terrarium/heracles/client/screens/mousemode/MouseModeButton.java:
```diff
@@ -1,16 +1,16 @@
 public class MouseModeButton extends SelectableImageButton {
 
-    private static MouseButtonType currentType = MouseButtonType.MOVE;
+    private static String currentType = "move";
```
What's the reason for making this not an enum?

JayemCeekay replied:
Will be fixed in next commit

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quests/SelectQuestWidget.java:
```diff
@@ -63,7 +63,7 @@ public SelectQuestWidget(int x, int y, int width, int height, QuestsWidget widge
             this.entry,
             quest -> {
-                quest.display().setTitle(s.isEmpty() ? null : Component.translatable(s));
+                quest.display().setTitle(s.isEmpty() ? null : Component.literal(s));
```
This should be kept as a translatable, a little further down if the title is translatable we will set the editbox to the key instead of the string

JayemCeekay replied:
This was related to the YabnOps.COMPRESSED issue I mentioned regarding the RestrictedEntityPredicate.CODEC in KillEntityQuestTask, I've reverted the change in this next commit.

JayemCeekay replied:
fixed in latest commits

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quests/SelectQuestWidget.java:
```diff
@@ -181,7 +181,7 @@ public void render(GuiGraphics graphics, int mouseX, int mouseY, float partialTi
         updateWidgets();
 
-        graphics.blitRepeating(AbstractQuestScreen.HEADING, this.x - 2, this.y, 2, this.height, 128, 0, 2, 256);
+        graphics.blit(AbstractQuestScreen.HEADING, this.x - 2, this.y, 2, this.height, 128, 0, 2, 256);
```
the change here changes how itd look since now its gonna stretch it instead of repeat, should be changed to a sprite so resource packs can choose how it will be rendered

JayemCeekay replied:
fixed in latest commits

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quest/editing/QuestMultiLineEditBox.java:
```diff
@@ -109,40 +110,47 @@ public boolean canClickText(double mouseX, double mouseY) {
+                    content.insertText("<obf>" + content.getSelectedText() + "</obf>");
+                    return true;
+                } else if (text.contains("#")) {
+                    content.insertText("<span style=\"color:" + text + "\">" + content.getSelectedText() + "</span>");
```
this should use the <red> elements to make it clearer to users

ThatGravyBoat replied:
elements like `<red>` and `<blue>`

JayemCeekay replied:
I was going to, but for some reason, and I believe this is upstream in minemark, it refuses to parse <dark_red> or anything with an _ in its name for colors. \<red\> works fine, but there must be something sanitizing the tag further upstream, so HermesText might not receive dark_red, but rather darkred. That's my best guess to what is happening.

ThatGravyBoat replied:
This may be a bug in Hermes tbh I think `_` may actually not be valid in actual HTML unlike before with its custom parser so I will likely have to change it to a `-`, ill go take a look at Hermes.

ThatGravyBoat replied:
if you update Hermes to 2.0.5 it will be fixed the underscores are now replayed with hyphens

JayemCeekay replied:
fixed in latest commits

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quest/editing/QuestMultiLineEditBox.java:
```diff
@@ -109,40 +110,47 @@ public boolean canClickText(double mouseX, double mouseY) {
+        BOLD("b"),
+        ITALIC("i"),
+        UNDERLINE("u"),
+        STRIKETHROUGH("strike"),
```
these should stay as the traditional markdown modifiers

JayemCeekay replied:
fixed in latest commits

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quest/QuestTagProvider.java:
```diff
@@ -6,12 +6,11 @@
 @AutoService(ElementExtension.class)
 public class QuestTagProvider implements ElementExtension {
 
-    @Override
```
The @Override should be kept

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/screens/quest/HeadingWidget.java:
```diff
@@ -12,16 +12,42 @@
+            case 3:
+                graphics.blitSprite(ResourceLocation.fromNamespaceAndPath(Heracles.MOD_ID, "widgets/heading_status_completed_claimed_other"), x + titleWidth + 6, y, width - titleWidth - 6, Minecraft.getInstance().font.lineHeight + 4);
+                break;
+        }
```
instead of ordinals this should use the enum itself aswell as using the modern switch syntax with arrows to avoid the break statements

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/api/tasks/defaults/KillEntityQuestTask.java:
```diff
@@ -55,14 +62,23 @@ public ResourceLocation id() {
         }
     }
+
+    public static final Codec<RestrictedEntityPredicate> CODEC = RecordCodecBuilder.create(instance -> instance.group(
```
This looks like a 1:1 copy of the RestrictedEntityPredicate.CODEC in resourceful lib I think it could use that

JayemCeekay replied:
resourceful lib version uses optionalFieldOf() instead of lenientOptionalFieldOf() which was causing crashes during packet decoding. This was an issue I encountered in various areas throughout the mod.

ThatGravyBoat replied:
what were the errors that occurred when sending it to the client? because now its just gonna silently fail which isnt the best situation

JayemCeekay replied:
[14:54:41] [Render thread/WARN] [minecraft/ClientCommonPacketListenerImpl]: Client disconnected with reason: Internal Exception: io.netty.handler.codec.DecoderException: Failed to decode packet 'serverbound/minecraft:custom_payload'

It is something to do with not having any default values to be read from "old" or something as far as I can tell in the create() in [KillEntityTaskSettings](https://github.com/JayemCeekay/Heracles/blob/1.21.1/common/src/main/java/earth/terrarium/heracles/api/client/settings/tasks/KillEntityTaskSettings.java).

ThatGravyBoat replied:
Is that all that it logs? should have some more details, my assumption is its throwing an error on one of the codec entries accessing the registry via registryops

JayemCeekay replied:
It seems to be an issue with using YabnOps.COMPRESSED. Switching to encoding and decoding using YabnOps.INSTANCE lets it decode without issues using the original RestrictedEntityPredicate.CODEC.

ThatGravyBoat replied:
The difference between COMPRESSED and INSTANCE is that it checks if the codec is keyable and if it is then it doesnt need to include object keys so on the client it can receieve the keys in the constant order so maybe there is a custom codec somewhere that is implementing keyable incorrectly since keys cant be duplicated

JayemCeekay replied:
It might be ComponentSerialization.CODEC in QuestDisplay. Changing it to ComponentSerialization.FLAT_CODEC lets everything load fine with YabnOps.COMPRESSED. However, it does change the way the title and subtitle sections are formatted. For example "title": "{\\"translate\\"\:\"abcd\\"}" vs the older "title": {"translate":"abcd"}

ThatGravyBoat replied:
That's okay since itemstacks changed between 1.20.x and 1.21.x so the format is already changed

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/mixins/client/ClientLevelMixin.java:
```diff
@@ -19,13 +21,15 @@ public class ClientLevelMixin {
         if (this.minecraft.gameMode.getPlayerMode() != GameType.CREATIVE) return;
         ItemStack itemStack = this.minecraft.player.getMainHandItem();
-        if (itemStack.is(ModItems.BARRIER.get())) {
+        if (itemStack.getItem().builtInRegistryHolder().is(BARRIER_ITEM_ID)) {
```
Why compare the ResourceLocations instead of comparing the items like before?

JayemCeekay replied:
reverted in latest commit

JayemCeekay committed:
changed inline usages of ResourceLocation.fromNameSpaceAndPath to constants and added a id() helper to Heracles.java.
Move usages of new NbtPredicate(new CompoundTag()) to a static utility in CustomizableQuestElement.java.
Changed currentType field of MouseButtonMode to use an enum instead of a string.
Changed a UI element in SelectQuestWidget to render using blitSprite()
Fixed QuestMultiLineEditBox formatting notation
Added missing @Override to QuestTagProvider
Changed switch statements in HeadingWidget and WidgetUtils to use enhanced switch statements and switch based off enum value rather than ordinal
Reverted unneeded change in ClientLevelMixin

JayemCeekay committed:
Fixed missing toast sprites rendering in QuestClaimedToast and QuestTutorialToast

MsRandom commented on common/src/main/java/earth/terrarium/heracles/api/client/WidgetUtils.java:
```diff
@@ -44,17 +46,26 @@ public static void drawSummaryBackground(GuiGraphics graphics, int x, int y, int
+            case LOCKED -> graphics.blitSprite(Heracles.id("widgets/summary_background_0"), x, y, width, height);
+            case IN_PROGRESS -> graphics.blitSprite(Heracles.id("widgets/summary_background_1"), x, y, width, height);
+            case COMPLETED -> graphics.blitSprite(Heracles.id("widgets/summary_background_2"), x, y, width, height);
+            case COMPLETED_CLAIMED -> graphics.blitSprite(Heracles.id("widgets/summary_background_3"), x, y, width, height);
```
The new sprites should include the enum case. ie `summery_background_locked`.

JayemCeekay committed:
Renamed summary_background_# sprites to use enum-based names such as summary_background_locked.png

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/api/rewards/defaults/ItemReward.java:
```diff
@@ -37,16 +38,16 @@ private static class Type implements QuestRewardType<ItemReward> {
-                ItemStackCodec.CODEC.fieldOf("item").forGetter(ItemReward::stack)
+                Codec.STRING.lenientOptionalFieldOf("title", "").forGetter(ItemReward::title),
+                QuestIcons.CODEC.lenientOptionalFieldOf("icon", ItemQuestIcon.AIR).forGetter(ItemReward::icon),
+                ItemStackCodec.CODEC.lenientOptionalFieldOf("item", Items.AIR.getDefaultInstance()).forGetter(ItemReward::stack)
```
this wasnt optional before

ThatGravyBoat reviewed (changes requested):
Most of the optionalFieldOfs that were changed to lenient should be changed back to normal optionalFieldOf and if any of the fields being switched back cause it to fail to encode or decode that should be looked into let me know if you need help with those

JayemCeekay committed:
Changed title fields in tasks and rewards to use lenient optional falling back to an empty string in order to prevent crashes when a title is not provided.
